src/rgw/rgw_qos.h

```diff
@@ -0,0 +1,246 @@
+#pragma once
+#include <chrono>
+#include <thread>
+#include <condition_variable>
+#include "rgw_common.h"
+
+#if !defined(dout_subsys)
+#define dout_subsys ceph_subsys_rgw
+#define def_dout_subsys
+#endif
+
+#define dout_context g_ceph_context
+#define fixed_float_rgw_ratelimit 1000
+class QosDatastruct {
+  enum {
+    RGW_QOS_OPS,
+    RGW_QOS_BW
+  };
+  struct qos_entry {
+    // 2 is the enum size
+    int64_t read[2] = {0};
+    int64_t write[2] = {0};
+    ceph::timespan ts;
+    bool first_run = true;
+    std::mutex ts_lock;
+    int64_t read_ops()
+    {
+      int64_t ret = read[RGW_QOS_OPS] / fixed_float_rgw_ratelimit;
+      return ret;
+    }
+    int64_t write_ops()
+    {
+      int64_t ret = write[RGW_QOS_OPS] / fixed_float_rgw_ratelimit;
+      return ret;
+    }
+    int64_t read_bytes()
+    {
+      int64_t ret = read[RGW_QOS_BW] / fixed_float_rgw_ratelimit;
+      return ret;
+    }
+    int64_t write_bytes()
+    {
+      int64_t ret = write[RGW_QOS_BW] / fixed_float_rgw_ratelimit;
+      return ret;
+    }
+  };
+  const size_t map_size = 2000000; // will create it with the closest upper prime number
+  std::shared_mutex insert_lock;
+  std::condition_variable *cv;
+  std::atomic_bool *replacing;
+  typedef std::unordered_map<std::string, qos_entry> hash_map;
+  hash_map qos_entries{map_size};
+  bool is_read_op(const char *method) {
```

Comment:
can be a static function, since it doesn't depend on any member variables. i also prefer std::string_view for comparisons:
```c++
static bool is_read_op(std::string_view method) {
  return method == "GET" || method == "HEAD";
}
```